### PR TITLE
Attendance(Augusto 30, 31) and HackerRank exercise of Collections

### PR DESCRIPTION
**Pull request by BigOsvaap**

### CODE REVIEW

File: Oswaldo Vazquez Aparicio/30 Agosto/hackerrank/hashtablesransomnote/HashTablesRansomNote.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+public class HashTablesRansomNote {
+
+    public static void main(String[] args) throws IOException {
+        BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(System.in));
+        String regex = "\\s+$";
+        String blankSpace = " ";
+        String[] firstMultipleInput = bufferedReader.readLine().replaceAll(regex, "").split(blankSpace);
+
+        int wordsInMagazine = Integer.parseInt(firstMultipleInput[0]);
```
Comment (luisod78) [inline Oswaldo Vazquez Aparicio/30 Agosto/hackerrank/hashtablesransomnote/HashTablesRansomNote.java:17]: wordsInMagazine??
wordsInNote??!!!

and these variables are here for.... ???

File: Oswaldo Vazquez Aparicio/30 Agosto/hackerrank/hashtablesransomnote/Result.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
+package hackerrank.hashtableransomnote;
+
+import java.util.HashMap;
+import java.util.List;
+
+class Result {
```
Comment (luisod78) [inline Oswaldo Vazquez Aparicio/30 Agosto/hackerrank/hashtablesransomnote/Result.java:6]: why no access modifier for class Result?

File: Oswaldo Vazquez Aparicio/30 Agosto/hackerrank/javasort/Student.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+    public Double getCgpa() {
+        return cgpa;
+    }
+
+    @Override
+    public int compareTo(Student student) {
+        int compareCgpa = cgpa.compareTo(student.getCgpa());
+
+        if (!areEquals(compareCgpa)) {
```
Comment (luisod78) [inline Oswaldo Vazquez Aparicio/30 Agosto/hackerrank/javasort/Student.java:31]: OMG!!!
I think you can refactor compareTo() so it would be more readable right?
areEquals() and  invertOrder() are really necessary??? really?

File: Oswaldo Vazquez Aparicio/30 Agosto/hackerrank/hashtablesransomnote/Result.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+                return;
+            }
+
+        }
+
+        System.out.println("Yes");
+
+    }
+
+    private static HashMap<String, Integer> createWordsFrequencyTable(List<String> words){
```
Comment (luisod78) [inline Oswaldo Vazquez Aparicio/30 Agosto/hackerrank/hashtablesransomnote/Result.java:28]: nice job!

**Review by luisod78 — CHANGES_REQUESTED** [review]

File: Oswaldo Vazquez Aparicio/30 Agosto/hackerrank/JavaArrayList.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
+package hackerrank;
+
+import java.util.ArrayList;
+import java.util.Scanner;
+
+public JavaArrayList {
+
+    private static final Scanner scanner = new Scanner(System.in);
+    private static List<List<Integer>> table;
```
Comment (luisod78) [inline Oswaldo Vazquez Aparicio/30 Agosto/hackerrank/JavaArrayList.java:9]: c'mon!!! 
You changed ArrayList<ArrayList<Integer>> for List<List<Integer>>
The thing is: why List<Integer> is not sufficient?

**Review by luisod78 — CHANGES_REQUESTED** [review]

File: Oswaldo Vazquez Aparicio/30 Agosto/hackerrank/javasort/Student.java [hunk 6]
```diff
@@ -28,12 +28,12 @@ public Double getCgpa() {
     public int compareTo(Student student) {
         int compareCgpa = cgpa.compareTo(student.getCgpa());
 
-        if (!areEquals(compareCgpa)) {
-            return invertOrder(compareCgpa);
+        if (compareCgpa != 0) {
```
Comment (luisod78) [inline Oswaldo Vazquez Aparicio/30 Agosto/hackerrank/javasort/Student.java:31]: looks better!

**Review by luisod78 — CHANGES_REQUESTED** [review]

File: Oswaldo Vazquez Aparicio/30 Agosto/hackerrank/JavaArrayList.java [hunk 7]
```diff
@@ -6,7 +6,7 @@
 public JavaArrayList {
 
     private static final Scanner scanner = new Scanner(System.in);
-    private static ArrayList<ArrayList<Integer>> table;
+    private static List<List<Integer>> table;
```
Comment (globant-bootcamp) [inline Oswaldo Vazquez Aparicio/30 Agosto/hackerrank/JavaArrayList.java:9]: When you variable is a Collection, it is good practice put name in singular and them the the type, example: tableListList, userSet, etc...

File: Oswaldo Vazquez Aparicio/30 Agosto/hackerrank/JavaArrayList.java [hunk 8]
```diff
@@ -24,19 +24,19 @@ public static void main(String[] args) {
 
     }
 
-    private static ArrayList<ArrayList<Integer>> createTable(int rowsNumber) {
-        ArrayList<ArrayList<Integer>> table = new ArrayList<>(rowsNumber);
+    private static List<List<Integer>> createTable(int rowsNumber) {
+        List<List<Integer>> table = new ArrayList<>(rowsNumber);
```
Comment (globant-bootcamp) [inline Oswaldo Vazquez Aparicio/30 Agosto/hackerrank/JavaArrayList.java:28]: When you variable is a Collection, it is good practice put name in singular and them the the type, example: tableListList, userSet, etc...

File: Oswaldo Vazquez Aparicio/30 Agosto/hackerrank/JavaArrayList.java [hunk 9]
```diff
@@ -24,19 +24,19 @@ public static void main(String[] args) {
...
         for (int rowIndex = 0; rowIndex < rowsNumber; rowIndex++) {
             int numbersInRow = scanner.nextInt();
-            ArrayList<Integer> row = createRow(numbersInRow);
+            List<Integer> row = createRow(numbersInRow);
```
Comment (globant-bootcamp) [inline Oswaldo Vazquez Aparicio/30 Agosto/hackerrank/JavaArrayList.java:31]: When you variable is a Collection, it is good practice put name in singular and them the the type, example: tableListList, userSet, etc...

File: Oswaldo Vazquez Aparicio/30 Agosto/hackerrank/JavaArrayList.java [hunk 10]
```diff
@@ -24,19 +24,19 @@ public static void main(String[] args) {
...
             table.add(row);
         }
         scanner.nextLine();
         return table;
     }
 
-    private static ArrayList<Integer> createRow(int numbersInRow) {
-        ArrayList<Integer> row = new ArrayList<>(numbersInRow);
+    private static List<Integer> createRow(int numbersInRow) {
+        List<Integer> row = new ArrayList<>(numbersInRow);
```
Comment (globant-bootcamp) [inline Oswaldo Vazquez Aparicio/30 Agosto/hackerrank/JavaArrayList.java:39]: When you variable is a Collection, it is good practice put name in singular and them the the type, example: tableListList, userSet, etc...

File: Oswaldo Vazquez Aparicio/30 Agosto/hackerrank/JavaArrayList.java [hunk 11]
```diff
@@ -59,7 +59,7 @@ private static void performQuery(int lineNumber, int position) {
 
         if (lineNumber <= linesNumber) {
 
-            ArrayList<Integer> line = table.get(lineNumber - 1);
+            List<Integer> line = table.get(lineNumber - 1);
```
Comment (globant-bootcamp) [inline Oswaldo Vazquez Aparicio/30 Agosto/hackerrank/JavaArrayList.java:62]: When you variable is a Collection, it is good practice put name in singular and them the the type, example: tableListList, userSet, etc...

File: Oswaldo Vazquez Aparicio/30 Agosto/hackerrank/JavaHashSet.java [hunk 12]
```diff
@@ -10,7 +11,7 @@ public static void main(String[] args) {
         Scanner scanner = new Scanner(System.in);
         int pairsNumber = scanner.nextInt();
 
-        HashSet<String> pairs = new HashSet<>();
+        Set<String> pairs = new HashSet<>();
```
Comment (globant-bootcamp) [inline Oswaldo Vazquez Aparicio/30 Agosto/hackerrank/JavaHashSet.java:14]: When you variable is a Collection, it is good practice put name in singular and them the the type, example: tableListList, userSet, etc...

File: Oswaldo Vazquez Aparicio/30 Agosto/hackerrank/JavaList.java [hunk 13]
```diff
@@ -2,11 +2,12 @@
 
 import java.util.LinkedList;
 import java.util.Scanner;
+import java.util.List;
 
 public class JavaList {
 
     private static final Scanner scanner = new Scanner(System.in);
-    private static final LinkedList<Integer> list = new LinkedList<>();
+    private static final List<Integer> list = new LinkedList<>();
```
Comment (globant-bootcamp) [inline Oswaldo Vazquez Aparicio/30 Agosto/hackerrank/JavaList.java:10]: When you variable is a Collection, it is good practice put name in singular and them the the type, example: tableListList, userSet, etc...

File: Oswaldo Vazquez Aparicio/30 Agosto/hackerrank/JavaMap.java [hunk 14]
```diff
@@ -1,19 +1,20 @@
...
 
 import java.util.HashMap;
+import java.util.Map;
 import java.util.Scanner;
 
 public class JavaMap {
 
     private static final Scanner scanner = new Scanner(System.in);
-    private static HashMap<String, Integer> phoneBook;
+    private static Map<String, Integer> phoneBook = new HashMap<>();
```
Comment (globant-bootcamp) [inline Oswaldo Vazquez Aparicio/30 Agosto/hackerrank/JavaMap.java:10]: When you variable is a Collection, it is good practice put name in singular and them the the type, example: tableListList, userSet, etc...

File: Oswaldo Vazquez Aparicio/30 Agosto/hackerrank/hashtablesransomnote/HashTablesRansomNote.java [hunk 15]
```diff
@@ -14,10 +14,6 @@ public static void main(String[] args) throws IOException {
         String blankSpace = " ";
         String[] firstMultipleInput = bufferedReader.readLine().replaceAll(regex, "").split(blankSpace);
 
-        int wordsInMagazine = Integer.parseInt(firstMultipleInput[0]);
-
-        int wordsInNote = Integer.parseInt(firstMultipleInput[1]);
-
         List<String> magazine = Arrays.asList(bufferedReader.readLine().replaceAll(regex, "").split(blankSpace));
```
Comment (globant-bootcamp) [inline Oswaldo Vazquez Aparicio/30 Agosto/hackerrank/hashtablesransomnote/HashTablesRansomNote.java:17]: When you variable is a Collection, it is good practice put name in singular and them the the type, example: tableListList, userSet, etc...

File: Oswaldo Vazquez Aparicio/30 Agosto/hackerrank/hashtablesransomnote/HashTablesRansomNote.java [hunk 16]
```diff
@@ -14,10 +14,6 @@ public static void main(String[] args) throws IOException {
...
 
         List<String> note = Arrays.asList(bufferedReader.readLine().replaceAll(regex, "").split(blankSpace));
```
Comment (globant-bootcamp) [inline Oswaldo Vazquez Aparicio/30 Agosto/hackerrank/hashtablesransomnote/HashTablesRansomNote.java:19]: When you variable is a Collection, it is good practice put name in singular and them the the type, example: tableListList, userSet, etc...

File: Oswaldo Vazquez Aparicio/30 Agosto/hackerrank/hashtablesransomnote/Result.java [hunk 17]
```diff
@@ -1,14 +1,15 @@
...
 import java.util.List;
 
-class Result {
+public class Result {
 
     public static void checkMagazine(List<String> magazine, List<String> note) {
 
-        HashMap<String, Integer> wordsFreqMagazine = createWordsFrequencyTable(magazine);
-        HashMap<String, Integer> wordsFreqNote = createWordsFrequencyTable(note);
+        Map<String, Integer> wordsFreqMagazine = createWordsFrequencyTable(magazine);
```
Comment (globant-bootcamp) [inline Oswaldo Vazquez Aparicio/30 Agosto/hackerrank/hashtablesransomnote/Result.java:11]: When you variable is a Collection, it is good practice put name in singular and them the the type, example: tableListList, userSet, etc...

File: Oswaldo Vazquez Aparicio/30 Agosto/hackerrank/hashtablesransomnote/Result.java [hunk 18]
```diff
@@ -1,14 +1,15 @@
...
+        Map<String, Integer> wordsFreqNote = createWordsFrequencyTable(note);
```
Comment (globant-bootcamp) [inline Oswaldo Vazquez Aparicio/30 Agosto/hackerrank/hashtablesransomnote/Result.java:12]: When you variable is a Collection, it is good practice put name in singular and them the the type, example: tableListList, userSet, etc...

File: Oswaldo Vazquez Aparicio/30 Agosto/hackerrank/hashtablesransomnote/Result.java [hunk 19]
```diff
@@ -25,8 +26,8 @@ public static void checkMagazine(List<String> magazine, List<String> note) {
 
     }
 
-    private static HashMap<String, Integer> createWordsFrequencyTable(List<String> words){
-        HashMap<String, Integer> frequencyTable = new HashMap<>();
+    private static Map<String, Integer> createWordsFrequencyTable(List<String> words){
+        Map<String, Integer> frequencyTable = new HashMap<>();
```
Comment (globant-bootcamp) [inline Oswaldo Vazquez Aparicio/30 Agosto/hackerrank/hashtablesransomnote/Result.java:30]: When you variable is a Collection, it is good practice put name in singular and them the the type, example: tableListList, userSet, etc...

**Review by globant-bootcamp — CHANGES_REQUESTED** [review]

File: Oswaldo Vazquez Aparicio/30 Agosto/hackerrank/JavaArrayList.java (hunk 5 above)
Comment (BigOsvaap) [reply Oswaldo Vazquez Aparicio/30 Agosto/hackerrank/JavaArrayList.java:9]: I changed to List instead of ArrayList because I read in the resource below that you shared, it said that is better to use the Interface, I thought that was my mistake hehe.

https://www.javaguides.net/2018/06/java-collection-framework-best-practices.html

I did my matryoshka of ArrayLists because I wanted to do something like the image on the exercise in HackerRank, in this case, I imagined the red rectangle like an ArrayList that would contain
other rectangles and these rectangles would contain a number, the exercise was about ArrayList so I decided to use an ArrayList instead of an Array, so I ended with ArrayList<ArrayList<Integer>>
then later I changed to List<List<Integer>> thinking that was the problem, Oops!.

1489168616-b25dd38013-arraylist

So basically that is how I come with this.

Now if the exercise wasn't about ArrayList, I could have used a two-dimension Array something like Integer[][] but the rows needed to be dynamic, so an Array wouldn't do the trick, so I think I could use an Array of ArrayList, in this case, something like this "ArrayList[] tableListArray", should I change it? Thanks for the feedback.

Comment (luisod78) [reply Oswaldo Vazquez Aparicio/30 Agosto/hackerrank/JavaArrayList.java]: OK

**Review by globant-bootcamp — APPROVED** [review]
